File: ChartsDemo/Classes/Demos/SinusBarChartViewController.m
//
//  SinusBarChartViewController.m
//  ChartsDemo
//
//  Created by Daniel Cohen Gindi on 17/3/15.
//
//  Copyright 2015 Daniel Cohen Gindi & Philipp Jahoda
//  A port of MPAndroidChart for iOS
//  Licensed under Apache License 2.0
//
//  https://github.com/danielgindi/Charts
//

#import "SinusBarChartViewController.h"
#import "ChartsDemo-Swift.h"

@interface SinusBarChartViewController () <ChartViewDelegate>

@property (nonatomic, strong) IBOutlet BarChartView *chartView;
@property (nonatomic, strong) IBOutlet UISlider *sliderX;
@property (nonatomic, strong) IBOutlet UITextField *sliderTextX;

@end

@implementation SinusBarChartViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"Sinus Bar Chart";
    
    self.options = @[
                     @{@"key": @"toggleValues", @"label": @"Toggle Values"},
                     @{@"key": @"toggleHighlight", @"label": @"Toggle Highlight"},
                     @{@"key": @"toggleHighlightArrow", @"label": @"Toggle Highlight Arrow"},
                     @{@"key": @"animateX", @"label": @"Animate X"},
                     @{@"key": @"animateY", @"label": @"Animate Y"},
                     @{@"key": @"animateXY", @"label": @"Animate XY"},
                     @{@"key": @"saveToGallery", @"label": @"Save to Camera Roll"},
                     @{@"key": @"togglePinchZoom", @"label": @"Toggle PinchZoom"},
                     @{@"key": @"toggleAutoScaleMinMax", @"label": @"Toggle auto scale min/max"},
                     @{@"key": @"toggleData", @"label": @"Toggle Data"},
                     ];
    
    _chartView.delegate = self;
    
    _chartView.descriptionText = @"";
    _chartView.noDataTextDescription = @"You need to provide data for the chart.";
    
    _chartView.drawBarShadowEnabled = NO;
    _chartView.drawValueAboveBarEnabled = YES;
    _chartView.maxVisibleValueCount = 60;
    _chartView.pinchZoomEnabled = NO;
    _chartView.drawGridBackgroundEnabled = NO;
    
    ChartXAxis *xAxis = _chartView.xAxis;
    xAxis.labelPosition = XAxisLabelPositionBottom;
    xAxis.labelFont = [UIFont fontWithName:@"HelveticaNeue-Light" size:10.f];
    xAxis.drawGridLinesEnabled = NO;
    xAxis.enabled = NO;
    
    ChartYAxis *leftAxis = _chartView.leftAxis;
    leftAxis.labelFont = [UIFont fontWithName:@"HelveticaNeue-Light" size:10.f];
    leftAxis.labelCount = 6;
    leftAxis.axisMinValue = -2.5;
    leftAxis.axisMaxValue = 2.5;
    leftAxis.granularityEnabled = true;
    leftAxis.granularity = 0.1;
    
    ChartYAxis *rightAxis = _chartView.rightAxis;
    rightAxis.drawGridLinesEnabled = NO;
    rightAxis.labelFont = [UIFont fontWithName:@"HelveticaNeue-Light" size:10.f];
    rightAxis.labelCount = 6;
    rightAxis.axisMinValue = -2.5;
    rightAxis.axisMaxValue = 2.5;
    rightAxis.granularity = 0.1;
        
    ChartLegend *l = _chartView.legend;
    l.position = ChartLegendPositionBelowChartLeft;
    l.form = ChartLegendFormSquare;
    l.formSize = 9.0;
    l.font = [UIFont systemFontOfSize:11.f];
    l.xEntrySpace = 4.0;
    
    _sliderX.value = 150.0;
    [self slidersValueChanged:nil];
    
    [_chartView animateWithXAxisDuration:2.0 yAxisDuration:2.0];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)updateChartData
{
    if (self.shouldHideData)
    {
        _chartView.data = nil;
        return;
    }
    
    [self setDataCount:(_sliderX.value)];
}

- (void)setDataCount:(int)count
{
    NSMutableArray *xVals = [[NSMutableArray alloc] init];
    NSMutableArray *entries = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < count; i++)
    {
        [xVals addObject:[@(i) stringValue]];
        [entries addObject:[[BarChartDataEntry alloc] initWithValue:sinf(M_PI * (i % 128) / 64.0) xIndex:i]];
    }
    
    BarChartDataSet *set = nil;
    if (_chartView.data.dataSetCount > 0)
    {
        set = (BarChartDataSet *)_chartView.data.dataSets[0];
        set.yVals = entries;
        _chartView.data.xValsObjc = xVals;
        [_chartView notifyDataSetChanged];
    }
    else
    {
        set = [[BarChartDataSet alloc] initWithYVals:entries label:@"Sinus Function"];
        set.barSpace = 0.4;
        [set setColor:[UIColor colorWithRed:240/255.f green:120/255.f blue:124/255.f alpha:1.f]];
        
        BarChartData *data = [[BarChartData alloc] initWithXVals:xVals dataSet:set];
        [data setValueFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:10.f]];
        [data setDrawValues:NO];
        
        _chartView.data = data;
    }
}

- (void)optionTapped:(NSString *)key
{
    [super handleOption:key forChartView:_chartView];
}

#pragma mark - Actions

- (IBAction)slidersValueChanged:(id)sender
{
    _sliderTextX.text = [@((int)_sliderX.value + 1) stringValue];
    
    [self updateChartData];
}

#pragma mark - ChartViewDelegate

- (void)chartValueSelected:(ChartViewBase * __nonnull)chartView entry:(ChartDataEntry * __nonnull)entry dataSetIndex:(NSInteger)dataSetIndex highlight:(ChartHighlight * __nonnull)highlight
{
    NSLog(@"chartValueSelected");
}

- (void)chartValueNothingSelected:(ChartViewBase * __nonnull)chartView
{
    NSLog(@"chartValueNothingSelected");
}

@end

File: ChartsDemo-OSX/ChartsDemo-OSX/Demos/SinusBarChartViewController.swift
//
//  FeedItem.swift
//  ChartsDemo-OSX
//
//  Copyright 2015 Daniel Cohen Gindi & Philipp Jahoda
//  Copyright © 2017 thierry Hentic.
//  A port of MPAndroidChart for iOS
//  Licensed under Apache License 2.0
//
//  https://github.com/danielgindi/ios-charts

import Foundation
import Cocoa
import Charts



open class SinusBarChartViewController: DemoBaseViewController
{
    @IBOutlet var chartView: BarChartView!
    
    @IBOutlet var mainMenu: NSMenu!
    
    @IBOutlet weak var sliderX: NSSlider!
    @IBOutlet weak var sliderTextX: NSTextField!
    
    override open func viewDidAppear()
    {
        super.viewDidAppear()
        view.window!.title = "Sinus Bar Chart"
    }

    override open func viewWillAppear()
    {
        chartView.animate(xAxisDuration: 1.0, yAxisDuration: 1.0)
    }

    
    override open func viewDidLoad()
    {
        super.viewDidLoad()
        let options = [["label": "Zoom In"],
                       ["label": "Zoom out"],
                       ["label": "Reset Zoom"],
                       ["label": "separator"],
                       ["label": "Toggle Values"],
                       
                       ["label": "begin insert"],
                       ["label": "linear"],
                       ["label": "separator"],
                       
                       ["label": "easeInQuad"],
                       ["label": "easeOutQuad"],
                       ["label": "easeInOutQuad"],
                       ["label": "separator"],
                       
                       ["label": "easeInCubic"],
                       ["label": "easeOutCubic"],
                       ["label": "easeInOutCubic"],
                       ["label": "separator"],
                       
                       ["label": "easeInQuart"],
                       ["label": "easeOutQuart"],
                       ["label": "easeInOutQuart"],
                       ["label": "separator"],
                       
                       ["label": "easeInQuint"],
                       ["label": "easeOutQuint"],
                       ["label": "easeInOutQuint"],
                       ["label": "separator"],
                       
                       ["label": "easeInSine"],
                       ["label": "easeOutSine"],
                       ["label": "easeInOutSine"],
                       ["label": "separator"],
                       
                       ["label": "easeInExpo"],
                       ["label": "easeOutExpo"],
                       ["label": "easeInOutExpo"],
                       ["label": "separator"],
                       
                       ["label": "easeInCirc"],
                       ["label": "easeOutCirc"],
                       ["label": "easeInOutCirc"],
                       ["label": "separator"],
                       
                       ["label": "easeInElastic"],
                       ["label": "easeOutElastic"],
                       ["label": "easeInOutElastic"],
                       ["label": "separator"],
                       
                       ["label": "easeInBack"],
                       ["label": "easeOutBack"],
                       ["label": "easeInOutBack"],
                       ["label": "separator"],
                       
                       ["label": "easeInBounce"],
                       ["label": "easeOutBounce"],
                       ["label": "easeInOutBounce"],
                       
                       ["label": "end insert"],
                       ["label": "Animate ..."],
                       
                       ["label": "Save to Camera Roll"],
                       ["label": "Toggle PinchZoom"],
                       ["label": "Toggle auto scale min/max"],
                       ["label": "Toggle Data"]]
       
        let subMenu = NSMenu()
        mainMenu.removeAllItems()
        subMenu.removeAllItems()
        var subMenuEnabled = false
        var addSubMenuEnable = false
        for option in options
        {
            switch (option["label"]!)
            {
            case "separator" :
                if subMenuEnabled == false && addSubMenuEnable == false
                {
                    mainMenu.addItem(NSMenuItem.separator())
                }
                else
                {
                    subMenu.addItem(NSMenuItem.separator())
                }
                
            case "begin insert" :
                subMenu.removeAllItems()
                subMenuEnabled = true
                
            case "end insert" :
                subMenuEnabled =  false
                addSubMenuEnable =  true
                
            default:
                if subMenuEnabled == false && addSubMenuEnable == false
                {
                    mainMenu.addItem(withTitle: option["label"]!, action: #selector(optionTapped(sender:)), keyEquivalent: "")
                }
                else
                {
                    if subMenuEnabled == true && addSubMenuEnable == false
                    {
                        subMenu.addItem(withTitle: option["label"]!, action: #selector(optionTapped(sender:)), keyEquivalent: "")
                    }
                    else
                    {
                        addSubMenuEnable =  false
                        let subMenuItem = NSMenuItem(title: option["label"]!, action: nil, keyEquivalent: "")
                        mainMenu.addItem(subMenuItem)
                        mainMenu.setSubmenu(subMenu, for: subMenuItem)
                    }
                }
            }
        }
        for item in mainMenu.items
        {
            item.target = self
        }
        
        // MARK: General
        chartView.drawBarShadowEnabled = false
        chartView.drawValueAboveBarEnabled = true
        chartView.maxVisibleCount = 60
        chartView.pinchZoomEnabled = false
        chartView.drawGridBackgroundEnabled = false
        chartView.drawBordersEnabled = true
        
        // MARK: xAxis
        let xAxis = chartView.xAxis
        xAxis.labelPosition = .bottom
        xAxis.labelFont = NSUIFont(name: "HelveticaNeue-Light", size: CGFloat(10.0))!
        xAxis.drawGridLinesEnabled = false
        xAxis.enabled = false
        
         // MARK: leftAxis
        let leftAxis = chartView.leftAxis
        leftAxis.labelFont = NSUIFont(name: "HelveticaNeue-Light", size: CGFloat(10.0))!
        leftAxis.labelCount = 6
        leftAxis.axisMinimum = -2.5
        leftAxis.axisMaximum = 2.5
        leftAxis.granularityEnabled = true
        leftAxis.granularity = 0.1
        
        // MARK: rightAxis
        let rightAxis = chartView.rightAxis
        rightAxis.drawGridLinesEnabled = false
        rightAxis.labelFont = NSUIFont(name: "HelveticaNeue-Light", size: CGFloat(10.0))!
        rightAxis.labelCount = 6
        rightAxis.axisMinimum = -2.5
        rightAxis.axisMaximum = 2.5
        rightAxis.granularity = 0.1
        
        // MARK: legend
        let legend = chartView.legend
        legend.horizontalAlignment = .left
        legend.verticalAlignment = .bottom
        legend.orientation = .horizontal
        legend.drawInside = false
        legend.form = .square
        legend.formSize = 9.0
        legend.font = NSUIFont.systemFont(ofSize: CGFloat(11.0))
        legend.xEntrySpace = 4.0
        
        // MARK: description
        chartView.chartDescription?.enabled = false
        
        sliderX.doubleValue  = 150.0
        slidersValueChanged(sliderX)
        
        setDataCount(150)
    }
    
    func updateChartData()
    {
        setDataCount  (Int(sliderX.intValue))
    }
    
    func setDataCount(_ count: Int) {
        
        // MARK: BarChartDataEntry
       var entries = [BarChartDataEntry]()
        for i in 0..<count
        {
            entries.append(BarChartDataEntry(x: Double(i), y: Double(sinf(Float(.pi * Double(i % 128) / 64.0)))))
        }
        
        // MARK: BarChartDataSet
       var set = BarChartDataSet()
        if chartView.data != nil
        {
            set = (chartView.data?.dataSets[0] as? BarChartDataSet)!
            set.values = entries
            
            chartView.data?.notifyDataChanged()
            chartView.notifyDataSetChanged()
        }
        else
        {
            set = BarChartDataSet(values: entries, label: "Sinus Function")
            set.colors = [NSUIColor(red: CGFloat(240 / 255.0), green: CGFloat(120 / 255.0), blue: CGFloat(124 / 255.0), alpha: 1.0)]
        }
        set.valueFont = NSUIFont(name: "HelveticaNeue-Light", size: CGFloat(10.0))!
        set.drawValuesEnabled = false
        set.barBorderWidth = 0.1
        
        // MARK: BarChartData
        let data = BarChartData(dataSet: set)
        chartView.data = data
    }
    
    func optionTapped( sender: NSMenuItem)
    {
        switch (sender.title)
        {
         default:
            super.toggle(sender.title, chartView: chartView)
        }
    }
    
    @IBAction func slidersValueChanged(_ sender: AnyObject)
    {
        sliderTextX.stringValue =  String(Int( sliderX.intValue))
        updateChartData()
    }
    
    @IBAction func zoomAll(_ sender: AnyObject)
    {
        chartView.fitScreen()
    }
    
    @IBAction func zoomIn(_ sender: AnyObject)
    {
        chartView.zoomToCenter(scaleX: 1.5, scaleY: 1)
    }
    
    @IBAction func zoomOut(_ sender: AnyObject)
    {
        chartView.zoomToCenter(scaleX: 2/3, scaleY: 1)
    }

}


